### PR TITLE
Feature/improve output json 2

### LINKED ISSUE #473: Be more precise on the iteration duration that is printed in the json output file
Currently, a run of Xpansion produces a `out.json` file that gathers information on the whole study and on each iteration. This file contains a field `duration` for each iteration and a field `duration` related to the whole study.

More precision and information should be added to understand what is measured by these timers:
  - we could rename the field of the global timer `run_duration`
  - the iteration timer corresponds in fact only to the duration of the master resolution, it should be renamed `master_duration`
  - the time of subproblems resolution should also be added, in a field named `subproblem_duration` for instance

### PR DESCRIPTION
**Pull request by a-zakir**
* work on durations fields in output file out.json

### CODE REVIEW

**Comment by a-zakir** [comment]
close #473

**Review by JasonMarechal25 — APPROVED** [review]

File: src/cpp/benders/benders_core/include/core/ILogger.h [hunk 1]
```diff
@@ -70,7 +71,8 @@ struct LogData {
            relative_gap == lhs.relative_gap &&
            max_iterations == lhs.max_iterations &&
            benders_elapsed_time == lhs.benders_elapsed_time &&
-           master_time == lhs.master_time;
+           master_time == lhs.master_time &&
+           subproblem_time == lhs.subproblem_time;
```
Comment (flomnes) [inline src/cpp/benders/benders_core/include/core/ILogger.h:75]: The compiler can generate this operator for us with the exact same behavior : compare all data members. I counted 16 data members and 16 `==` comparisons. Unless you want to ignore some data members in the comparison, I suggest you remove lines 64-75.

To add some weight to my argument, `master_time` was previously forgotten, you added it in this commit.

File: src/cpp/benders/benders_core/BendersBase.cpp [hunk 2]
```diff
@@ -251,6 +251,7 @@ void BendersBase::update_trace() {
   _trace[_data.it - 1]->_min_invest = std::make_shared<Point>(_data.min_invest);
   _trace[_data.it - 1]->_deleted_cut = _data.deletedcut;
   _trace[_data.it - 1]->_time = _data.timer_master;
+  _trace[_data.it - 1]->_subproblem_duration = _data.subproblem_timers;
   _trace[_data.it - 1]->_nbasis = _data.nbasis;
   _trace[_data.it - 1]->_invest_cost = _data.invest_cost;
   _trace[_data.it - 1]->_operational_cost = _data.subproblem_cost;
```
Comment (flomnes) [inline src/cpp/benders/benders_core/BendersBase.cpp]: Add reference to `_trace[_data.it - 1]` to improve readability. Something like `auto latest_trace& = _trace[_data.it - 1]`

File: src/cpp/benders/benders_core/include/BendersStructsDatas.h [hunk 3]
```diff
@@ -47,6 +47,7 @@ class WorkerMasterData {
   int _deleted_cut;
   int _nbasis;
   double _time;
+  double _subproblem_duration;
```
Comment (flomnes) [inline src/cpp/benders/benders_core/include/BendersStructsDatas.h]: Rename `_time` to `_master_time` to clarify ?

File: src/cpp/benders/output/include/OutputWriter.h [hunk 4]
```diff
@@ -28,6 +30,7 @@ struct CandidateData {
 typedef std::vector<CandidateData> CandidatesVec;
 struct Iteration {
   double time;
+  double subproblem_duration;
```
Comment (flomnes) [inline src/cpp/benders/output/include/OutputWriter.h]: `time` is inaccurate, consider renaming

File: src/cpp/benders/benders_core/LastIterationReader.cpp [hunk 5]
```diff
@@ -51,7 +51,10 @@ LogData LastIterationReader::GetIterationData(
       last_iteration_file_content_[iteration_name]["max_iterations"].asInt(),
       last_iteration_file_content_[iteration_name]["benders_elapsed_time"]
           .asDouble(),
-      last_iteration_file_content_[iteration_name]["duration"].asDouble()
+      last_iteration_file_content_[iteration_name]["master_duration"]
+          .asDouble(),
+      last_iteration_file_content_[iteration_name]["subproblem_duration"]
+          .asDouble()
```
Comment (flomnes) [inline src/cpp/benders/benders_core/LastIterationReader.cpp:57]: Can you document those ? I can't see the difference between benders and master

File: src/cpp/benders/benders_core/include/core/ILogger.h (hunk 1 above)
Comment (a-zakir) [reply src/cpp/benders/benders_core/include/core/ILogger.h:75]: in c++ default `operator==()` is not available for c++std17.
`master_time` existed in the previous version, what's added is the new field `subproblem_time`.
This operator is used only for test purpose.

File: src/cpp/benders/benders_core/LastIterationReader.cpp (hunk 5 above)
Comment (a-zakir) [reply src/cpp/benders/benders_core/LastIterationReader.cpp:57]: master_duration is the duration of the master resolution at each iteration.
benders_elapsed_time is the total duration of all iterations (~sum(master_time+subproblem_time)).

File: src/cpp/benders/benders_core/include/core/ILogger.h (hunk 1 above)
Comment (flomnes) [reply src/cpp/benders/benders_core/include/core/ILogger.h:75]: You're right. C++20 introduced default `operator==` though. Why do you need this comparison operator ? Tests ?

Comment (a-zakir) [reply src/cpp/benders/benders_core/include/core/ILogger.h:75]: yep only for tests.

**Review by flomnes — APPROVED** [review]

File: src/cpp/benders/output/include/OutputWriter.h (hunk 4 above)
Comment (a-zakir) [reply src/cpp/benders/output/include/OutputWriter.h]: time is renamed master_duration

File: src/cpp/benders/benders_core/include/BendersStructsDatas.h (hunk 3 above)
Comment (a-zakir) [reply src/cpp/benders/benders_core/include/BendersStructsDatas.h]: _time is renamed _master_duration

**Pull request merged**